include/boost/iostreams/filter/zlib.hpp

```diff
@@ -347,7 +347,7 @@ bool zlib_compressor_impl<Alloc>::filter
     int result = xdeflate(flush ? zlib::finish : zlib::no_flush);
     after(src_begin, dest_begin, true);
     zlib_error::check BOOST_PREVENT_MACRO_SUBSTITUTION(result);
-    return result != zlib::stream_end; 
+    return result == zlib::okay; 
```

Comment:
OK, I agree with your analysis. I would still like to see, in place of `result == zlib::okay` the specific test for end-of-stream and buffer error in the form of `result != zlib::stream_end && result != zlib::buf_error` instead. The same goes for your second case below. I am being as conservative as I can be in this change so as to invoke only the particular consequences by which a Z_BUF_ERROR is now handled differently. In other words currently not reaching the end-of-stream returns 'true', and any other result returns 'false' while with your fix I want to have it explicitly changed so that not reaching the end-of-stream and not having a buffer error returns 'true', and any other result returns 'false'.